### PR TITLE
DScanner: check for assignment to auto ref parameters

### PR DESCRIPTION
**Pull request by wilzbach**
Yet another neat check.

### CODE REVIEW

File: std/algorithm/mutation.d [hunk 1]
```diff
@@ -2586,7 +2586,7 @@ Params:
     i1 = first index
     i2 = second index
 */
-void swapAt(R)(auto ref R r, size_t i1, size_t i2)
+void swapAt(R)(ref R r, size_t i1, size_t i2)
```
Comment (JackStouffer) [inline std/algorithm/mutation.d]: this is a breaking change

Comment (wilzbach) [inline std/algorithm/mutation.d]: Imho it's a bug fix, consider this example:

```d
import std.algorithm.mutation : swap;

void swapAtRef(R)(ref R r, size_t i1, size_t i2)
{
    static if (is(typeof(&r.swapAt)))
    {
        r.swapAt(i1, i2);
    }
    else static if (is(typeof(&r[i1])))
    {
        swap(r[i1], r[i2]);
    }
    else
    {
        if (i1 == i2) return;
        auto t1 = r.moveAt(i1);
        auto t2 = r.moveAt(i2);
        r[i2] = t1;
        r[i1] = t2;
    }
}

void swapAtValue(R)(R r, size_t i1, size_t i2)
{
    static if (is(typeof(&r.swapAt)))
    {
        r.swapAt(i1, i2);
    }
    else static if (is(typeof(&r[i1])))
    {
        swap(r[i1], r[i2]);
    }
    else
    {
        if (i1 == i2) return;
        auto t1 = r.moveAt(i1);
        auto t2 = r.moveAt(i2);
        r[i2] = t1;
        r[i1] = t2;
    }
}

void main(string[] args)
{
    import std.stdio;

    auto a = [1, 2, 3];
    a.swapAtRef(1, 2);
    writeln(a); // [1, 3, 2]

    auto b = [1, 2, 3];
    a.swapAtValue(1, 2);
    writeln(b); // [1, 2, 3]
}
```

Comment (JackStouffer) [inline std/algorithm/mutation.d]: Right, the current code fails silently for some ranges. But, there are ranges which give its elements by reference that could be part of a UFCS chain, and therefore don't have a address. Those will no longer work with this change.

**Comment by wilzbach** [comment]
Closing this as it's simply not worth the effort. If anyway wants to push for enabling more DScanner plugins, I am happy to help, but currently I simply don't have the time to maintain low-value work.

**Pull request closed without merging**